### DISCUSSION TITLE
How to install NERDCTL on existing IPFS environment?

### DISCUSSION THREAD

**Discussion by JvD007** [containerd/nerdctl#2717] [Q&A]
Hi All,

I am trying to install nerdctl on our IPFS private production environment. unfortunately it has not worked yet, what steps should you do to install nerdctl on an existing IPFS environment. 

With` -- install-ipfs `it does not work, then an ipfs service is created which conflicts with the existing installation.

We also have Bacalhau running on the same system and that should continue to work, the goal is to put containers on IPFS. 

We want to start using nerdctl, but we are stuck.

Any support is welcome

**Comment by AkihiroSuda** [comment]
It should just work, as long as `~/.ipfs/api` is present

```bash
echo "/ip4/<ip>/tcp/<rpc-port>" > ~/.ipfs/api
nerdctl pull ubuntu:20.04
nerdctl push ipfs://ubuntu:20.04
```

https://docs.ipfs.tech/install/command-line/
https://github.com/containerd/nerdctl/blob/main/docs/ipfs.md#using-nerdctl-with-ipfs

**Comment by JvD007** [comment]
After following the instructions we still don’t get it up and running

`$ containerd-rootless-setuptool.sh install`

and also tested with 

`$ containerd-rootless-setuptool.sh -- install-ipfs`

This is the message we got: 

```
nerdctl push ipfs://ubuntu:20.04
INFO[0000] pushing image "ubuntu:20.04" to IPFS         
INFO[0000] ensuring image contents                      
WARN[0000] ipfs push failed                              error="Post \"http://127.0.0.1:5001/api/v0/add?cid-version=1&pin=true\": dial tcp 127.0.0.1:5001: connect: connection refused"
FATA[0000] Post "http://127.0.0.1:5001/api/v0/add?cid-version=1&pin=true": dial tcp 127.0.0.1:5001: connect: connection refused 
```

IPFS setup is  in a Private setup (with swarm.key)

The IPFS_PATH is set to /ipfs-storage/ipfs

The ipfs.service is as follows:
```

[Unit] 

Description=IPFS Daemon 
After=network.target

[Service] 
User=odds
Environment=IPFS_PATH=/ipfs-storage/ipfs 
ExecStart=/usr/local/bin/ipfs daemon 
StandardOutput=journal 
Restart=on-failure 
KillSignal=SIGINT

[Install] 
WantedBy=multi-user.target

```
We have change API address to
`ipfs config Addresses.API /ip4/0.0.0.0/tcp/5001
`
We have also, used the local ip address of the server. We did also:

`echo "/ip4/192.168.122.204/tcp/5001" /ipfs-storage/ipfs/api
`
What is the best installation steps for getting nerdctl up-and-running on a running platform, I think we are missing something?

**Reply by AkihiroSuda** [reply]
Rootless mode has a different view of 127.0.0.1.
Probably `sudo nerdctl` should just work.

**Reply by JvD007** [reply]
You need to make some more steps before it works
```

$sudo nerdctl push ipfs://ubuntu:20.04

INFO[0000] pushing image "ubuntu:20.04" to IPFS         
INFO[0000] ensuring image contents                      
WARN[0000] failed to ensure the existence of image "ubuntu:20.04"  error="open /root/.ipfs/api: no such file or directory"
WARN[0000] ipfs push failed                              error="open /root/.ipfs/api: no such file or directory"
FATA[0000] open /root/.ipfs/api: no such file or directory 
```

Next step:

```
$sudo su
$cd /root

create .ipfs directory  and create the API file

$echo "/ip4/0.0.0.0/tcp/5001" > ~/.ipfs/api

$exit
```

What I see now you need to run everything with `sudo`

```
$ sudo nerdctl pull ubuntu:22.04

docker.io/library/ubuntu:22.04:                                                   resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:6042500cf4b44023ea1894effe7890666b0c5c7871ed83a97c36c76ae560bb9b:    done           |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:bbf3d1baa208b7649d1d0264ef7d522e1dc0deeeaaf6085bf8e4618867f03494: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:174c8c134b2a94b5bb0b37d9a2b6ba0663d82d23ebf62bd51f74a2fd457333da:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a486411936734b0d1d201c8a0ed8e9d449a64d5033fdc33411ec95bc26460efb:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 6.3 s                                                                    total:  28.2 M (4.5 MiB/s)     
```                                  
```

$ sudo nerdctl push ipfs://ubuntu:22.04

INFO[0000] pushing image "ubuntu:22.04" to IPFS         
INFO[0000] ensuring image contents                      
bafkreibpruqungbiqwpdvsf66ayx2ygz7earj7njnmldjbs3m3hgleq4lq
```

Not tested yet, if we can get it running on a second server.

Can we run this as the same way we can run Docker in  user mode?

**Reply by AkihiroSuda** [reply]
> echo "/ip4/192.168.122.204/tcp/5001" /ipfs-storage/ipfs/api 

Maybe you forgot to specify `>`  (file redirection)?
It might be also worth trying to symlink `/ipfs-storage/ipfs` to `~/.ipfs`

> Can we run this as the same way we can run Docker in user mode?

By "Docker in user mode" do you mean "nerdctl in rootless mode" ?
It should work, but you'll have to specify the IP explicitly in `~/.ipfs/api` to use your own IPFS installation.

**Comment by JvD007** [comment]
Oops, that really was a typo! I did it again and now it works

`$echo "/ip4/192.168.122.204/tcp/5001" > /ipfs-storage/ipfs/api `

Will continue testing!

**Comment by JvD007** [comment]
It was working, but after a reboot we see that the API file is changed to `/ip4/127.0.0.1/tcp/5001`

And if we now Pull and Push  container we got this message

```
$ nerdctl pull ubuntu:20.04

docker.io/library/ubuntu:20.04:                                                   resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:f2034e7195f61334e6caff6ecf2e965f92d11e888309065da85ff50c617732b8:    done           |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:bac6081102aae54ba4bcc714695b8f637e42768c7f376f374c428bab043ddc0f: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:f78909c2b360d866b3220655c0b079838258b8891a12ac25fc670f0cbb54229f:   done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 2.2 s                                                                    total:  3.8 Ki (1.7 KiB/s) 
```                                      

```
$ nerdctl push ipfs://ubuntu:20.04

INFO[0000] pushing image "ubuntu:20.04" to IPFS         
INFO[0000] ensuring image contents                      
WARN[0000] failed to ensure the existence of image "ubuntu:20.04"  error="no CID is recorded"
WARN[0000] ipfs push failed                              error="content digest sha256:527f5363b98e562da03d2e0bbf86fd7c34f487bffd9b27a3cf0a9afea2f0ee1f: not found"
FATA[0000] content digest sha256:527f5363b98e562da03d2e0bbf86fd7c34f487bffd9b27a3cf0a9afea2f0ee1f: not found 
```

Changing the API entry did not make any difference and the container is listed

```
$ nerdctl images

REPOSITORY    TAG      IMAGE ID        CREATED          PLATFORM       SIZE        BLOB SIZE
ubuntu        20.04    f2034e7195f6    4 minutes ago    linux/amd64    77.9 MiB    26.2 MiB
```

This had nothing todo with IPFS I think, where to look?

**Reply by AkihiroSuda** [reply]
Cc @ktock

**Comment by JvD007** [comment]
**Fix for the 127.0.0.1:5001: connect: connection refused**

If in the IPFS configuration file the API IP address is set to "API": "/ip4/0.0.0.0/tcp/5001", then the IP address in the ./ipfs/api ( in my situation /ipfs-storage/ipfs/api) will be set to /ip4/127.0.0.1/tcp/5001.

Then you get the following error message:

```
nerdctl push ipfs://fedora:rawhide

INFO[0000] moving image "fedora:rawhide" to IPFS       
INFO[0000] ensure content of image                      
WARN[0000] ipfs push failed error="Post "http://127.0.0.1:5001/api/v0/add?cid-version=1&pin=true: dial tcp 127.0.0.1:5001: connect: connection refused"
FATA[0000] Post "http://127.0.0.1:5001/api/v0/add?cid-version=1&pin=true": dial tcp 127.0.0.1:5001: connect: connection refused 
```

Only if the IP address in the IPFS config file is changed to "API": "/ip4/192.168.122.152/tcp/5001", then the IP address in the ./ipfs/api is directly changed to /ip4/192.168.122.152/tcp/5001. 

After changing the IPFS config file, you must restart the ipfs.service, only then will the change take effect.
